src/resources/gpio_esp32.cc

```diff
@@ -49,14 +49,45 @@ ResourcePool<int, -1> gpio_pins(
     32, 33, 34, 35, 36, 37, 38, 39
 );
 
+class GPIOResource : public EventQueueResource {
+public:
+  TAG(GPIOResource);
+
+  GPIOResource(ResourceGroup* group, int pin)
+      // GPIO resources share a queue, which is always on the event source, so pass null.
+      : EventQueueResource(group, null)
+      , _pin(pin) {}
+
+  int pin() const { return _pin; }
+
+  bool check_gpio(word pin) override;
+
+private:
+  int _pin;
+};
+
 class GPIOResourceGroup : public ResourceGroup {
  public:
   TAG(GPIOResourceGroup);
   explicit GPIOResourceGroup(Process* process)
-      : ResourceGroup(process, GPIOEventSource::instance()) {}
+      : ResourceGroup(process, EventQueueEventSource::instance()) {
+    queue = EventQueueEventSource::instance()->gpio_queue();
+  }
+
+  virtual void on_register_resource(Resource* r) {
```

Comment:
I would probably move the code for on_register and on_unregister out of the class definition.